core/src/main/java/org/apache/iceberg/view/ViewMetadataParser.java

```diff
@@ -0,0 +1,227 @@
+    Preconditions.checkArgument(
+        formatVersion <= ViewMetadata.SUPPORTED_VIEW_FORMAT_VERSION,
+        "Cannot read unsupported version %s",
+        formatVersion);
+
+    String location = JsonUtil.getString(LOCATION, node);
+    int currentVersionId = JsonUtil.getInt(CURRENT_VERSION_ID, node);
+    Map<String, String> properties = JsonUtil.getStringMap(PROPERTIES, node);
+
+    JsonNode versionsListNode = JsonUtil.get(VERSIONS, node);
+    Preconditions.checkArgument(
+        versionsListNode.isArray(), "Cannot parse versions from non-array: %s", versionsListNode);
+
+    List<ViewVersion> versions = Lists.newArrayListWithExpectedSize(versionsListNode.size());
+    ViewVersion currentVersion = null;
+    for (JsonNode versionNode : versionsListNode) {
+      ViewVersion version = ViewVersionParser.fromJson(versionNode);
+      if (version.versionId() == currentVersionId) {
+        currentVersion = version;
+      }
+
+      versions.add(version);
+    }
+
+    Preconditions.checkArgument(
+        currentVersion != null,
+        "Cannot find version with %s=%s from %s",
+        CURRENT_VERSION_ID,
+        currentVersionId,
+        VERSIONS);
+
+    JsonNode versionLogNode = JsonUtil.get(VERSION_LOG, node);
+    Preconditions.checkArgument(
+        versionLogNode.isArray(), "Cannot parse version-log from non-array: %s", versionLogNode);
+    List<ViewHistoryEntry> historyEntries =
+        Lists.newArrayListWithExpectedSize(versionLogNode.size());
+    Iterator<JsonNode> versionLogIterator = versionLogNode.elements();
+    while (versionLogIterator.hasNext()) {
+      historyEntries.add(ViewHistoryEntryParser.fromJson(versionLogIterator.next()));
+    }
+
+    List<Schema> schemas;
+    Integer currentSchemaId;
+    Schema currentSchema = null;
+    JsonNode schemaArray = node.get(SCHEMAS);
+
+    Preconditions.checkArgument(
+        schemaArray.isArray(), "Cannot parse schemas from non-array: %s", schemaArray);
+    currentSchemaId = JsonUtil.getInt(CURRENT_SCHEMA_ID, node);
+
+    ImmutableList.Builder<Schema> builder = ImmutableList.builder();
+    for (JsonNode schemaNode : schemaArray) {
+      Schema schema = SchemaParser.fromJson(schemaNode);
+      if (schema.schemaId() == currentSchemaId) {
+        currentSchema = schema;
+      }
+
+      builder.add(schema);
+    }
+
+    Preconditions.checkArgument(
+        currentSchema != null,
+        "Cannot find schema with %s=%s from %s",
+        CURRENT_SCHEMA_ID,
+        currentSchemaId,
+        SCHEMAS);
+
+    schemas = builder.build();
+
+    int numVersionsToKeep =
+        PropertyUtil.propertyAsInt(
+            properties,
+            ViewProperties.VERSION_HISTORY_SIZE,
+            ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+
+    Preconditions.checkArgument(
+        numVersionsToKeep >= 1, "%s must be positive", ViewProperties.VERSION_HISTORY_SIZE);
+
+    if (versions.size() > numVersionsToKeep) {
```

Comment:
Why do this here? I don't think it belongs in the parser.

Reply:
I agree that this shouldn't be here. The better option would be to put this into a special version of a `@Check` method that allows normalization. This would allow to retain the number of versions that are desired in `ViewMetadata` directly

Reply:
Agreed, @nastra fixed this in https://github.com/apache/iceberg/pull/7759/files#diff-037c118271eb576698f0829d876d5c49919eea4bd20e6b2f98b6072cfd0d4b08R133-R135 where all the validation is done in the special check method he referred to.